src/node_api.cc

```diff
@@ -3551,3 +3552,233 @@ napi_status napi_run_script(napi_env env,
+                    napi_threadsafe_function_default_call_js : call_js_cb;
+                ts_fn->hint = hint;
+                ts_fn->env = env;
+                ts_fn->max_queue_size = max_queue_size;
+                ts_fn->idle_running = false;
+
+                *result = reinterpret_cast<napi_threadsafe_function>(ts_fn);
+                return napi_clear_last_error(env);
+              }
+              if (max_queue_size != 0) {
+                uv_cond_destroy(&ts_fn->cond);
+              }
+            }
+            status = napi_delete_reference(env, ts_fn->ref);
+
+            // Although we have successfully deleted the reference, we have
+            // failed to create the thread-safe function;
+            if (status == napi_ok) {
+              status = napi_generic_failure;
+            }
+          }
+          uv_mutex_destroy(&ts_fn->mutex);
+        }
+        uv_close(reinterpret_cast<uv_handle_t*>(&ts_fn->async),
+            [] (uv_handle_t* handle) -> void {
+              delete handle;
+            });
+
+        // Prevent the thread-safe function from being deleted here, because the
+        // uv_close() callback above will delete it.
+        ts_fn = nullptr;
+      }
+      delete ts_fn;
+    }
+  }
+  return napi_set_last_error(env, status);
+}
+
+NAPI_EXTERN napi_status
+napi_call_threadsafe_function(napi_threadsafe_function async,
+                              void* data,
+                              bool* result) {
+  bool local_result = false;
+  napi_status status = napi_ok;
+  uv_mutex_lock(&async->mutex);
+  while (async->queue.size() >= async->max_queue_size &&
+      async->max_queue_size > 0) {
+    if (result != nullptr) {
+      goto done;
+    }
+    uv_cond_wait(&async->cond, &async->mutex);
+  }
+  status = uv_async_send(&async->async) == 0 ? napi_ok : napi_generic_failure;
+  if (status == napi_ok) {
+    async->queue.push(data);
+    local_result = true;
+  }
+done:
+  if (result != nullptr) {
+    *result = local_result;
+  }
+  uv_mutex_unlock(&async->mutex);
+  return status;
+}
+
+NAPI_EXTERN napi_status
+napi_delete_threadsafe_function(napi_env env,
+                                napi_threadsafe_function ts_fn) {
+  CHECK_ENV(env);
+  CHECK_ARG(env, ts_fn);
+
+  napi_status status = napi_delete_reference(env, ts_fn->ref);
+  if (status != napi_ok) {
+    return status;
+  }
+
+  uv_mutex_destroy(&ts_fn->mutex);
+  if (ts_fn->max_queue_size > 0) {
+    uv_cond_destroy(&ts_fn->cond);
+  }
+
+  uv_close(reinterpret_cast<uv_handle_t*>(&ts_fn->idle),
+      [] (uv_handle_t* handle) -> void {
+        napi_threadsafe_function ts_fn =
+            node::ContainerOf(&napi_threadsafe_function__::idle,
+              reinterpret_cast<uv_idle_t*>(handle));
+        uv_close(reinterpret_cast<uv_handle_t*>(&ts_fn->async),
+          [] (uv_handle_t* handle) -> void {
+            delete handle;
+          });
+      });
+
+  return napi_clear_last_error(env);
```

Comment:
I might be missing it - is there a place ts_fn gets deleted?

Reply:
`ts_fn` == `handle` in the innermost lambda. @bnoordhuis advised me to stop making the assumption throughout the code that &ts_fn->async == ts_fn, but I don't believe I can avoid making that assumption when deleting the handle, because all I get from libuv is a `uv_handle_t *handle` for the innermost lambda.

Reply:
Actually, I found a way of removing the assumption.